8335221: Some C2 intrinsics incorrectly assume that type argument is compile-time constant

vnkozlov opened the pull request:
Problems were found in some C2 intrinsics while testing Leyden Early Release.

[`inline_array_partition()`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/library_call.cpp#L5493) and [`inline_array_sort()`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/library_call.cpp#L5559) intrinsics incorrectly assume that `elementType` argument is always compile-time constant. Which is not true for Leyden (and in general) when constant folding optimization is switched off (or did not executed for particular type).

Note, other intrinsics are very careful about that. For example, see [`inline_Class_cast()`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/library_call.cpp#L3963) intrinsic.

Add similar checks to `inline_array_partition()` and `inline_array_sort()`.

An other problem is that `null_check()` (which adds new control nodes to graph) was executed in these intrinsics before bailouts from intrinsic generation. We hit next assert if bailout happens:
```
  assert(ctrl == kit.control(), "Control flow was added although the intrinsic bailed out");
```

I moved `null_check()` after bailouts.

An other issue with `null_check()` is it checks for `null` first argument which is `Class`. It can't be `null` because corresponding Java methods are `private` and each call site passed `<primitive_type>.class` as first argument.
For example: [DualPivotQuicksort.java#L260](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/DualPivotQuicksort.java#L260)
On other hand there is no null check for passed array (second argument). I changed `null_check()` for array.

These 2 intrinsics were added by [JDK-8309130](https://bugs.openjdk.org/browse/JDK-8309130) in JDK 22.

Testing tier1-6, hs-stress, hs-xcomp

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335221](https://bugs.openjdk.org/browse/JDK-8335221): Some C2 intrinsics incorrectly assume that type argument is compile-time constant (**Bug** - P3)


### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [c370b07a](https://git.openjdk.org/jdk/pull/19918/files/c370b07ab4d3cdeca9c91dbfa4ce0fda6973c0b0)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [8ab7034a](https://git.openjdk.org/jdk/pull/19918/files/8ab7034a335db2f80cc0bdc4fa3c3cdc3b56fccd)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19918/head:pull/19918` \
`$ git checkout pull/19918`

Update a local copy of the PR: \
`$ git checkout pull/19918` \
`$ git pull https://git.openjdk.org/jdk.git pull/19918/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19918`

View PR using the GUI difftool: \
`$ git pr show -t 19918`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19918.diff">https://git.openjdk.org/jdk/pull/19918.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19918#issuecomment-2192927692)

rwestrel reviewed (approved):
Looks good to me.

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5502,29 +5502,44 @@ bool LibraryCallKit::inline_array_partition() {
+  if (!Matcher::supports_simd_sort(bt)) {
+    return false;
+  }
+  address stubAddr = nullptr;
+  stubAddr = StubRoutines::select_array_partition_function();
```
Can be merged:
```suggestion
  address stubAddr = StubRoutines::select_array_partition_function();
```

vnkozlov replied:
done

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5502,29 +5502,44 @@ bool LibraryCallKit::inline_array_partition() {
+
+  null_check(obj);
+  // If obj is dead, only null-path is taken.
+  if (stopped())  return true;
```
Can you add braces here?
```suggestion
  if (stopped()) {
    return true;
  }
```

vnkozlov replied:
done

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5502,29 +5502,44 @@ bool LibraryCallKit::inline_array_partition() {
 
   Node* pivotIndices = nullptr;
 
+  const TypeInstPtr* elem_klass = gvn().type(elementType)->isa_instptr();
+  if (elem_klass == nullptr) {
+    return false;  // dead path (elementType->is_top()).
+  }
+  if (obj == nullptr || obj->is_top()) {
+    return false;  // dead path
+  }
+  // java_mirror_type() returns non-null for compile-time Class constants only.
+  ciType* elem_type = elem_klass->java_mirror_type();
+  if (elem_type == nullptr) {
+    return false;
+  }
+  BasicType bt = elem_type->basic_type();
+  // Disable the intrinsic if the CPU does not support SIMD sort
+  if (!Matcher::supports_simd_sort(bt)) {
+    return false;
+  }
+  address stubAddr = nullptr;
+  stubAddr = StubRoutines::select_array_partition_function();
+  // stub not loaded
+  if (stubAddr == nullptr) {
+    return false;
+  }
+  // get the address of the array
+  const TypeAryPtr* obj_t = _gvn.type(obj)->isa_aryptr();
+  if (obj_t == nullptr || obj_t->elem() == Type::BOTTOM ) {
+    return false; // failed input validation
+  }
```
The bailout code looks almost identical to the one in `inline_array_sort()`. Can it somehow be shared?

vnkozlov replied:
Good idea. I will try.

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5583,6 +5608,11 @@ bool LibraryCallKit::inline_array_sort() {
+
+  null_check(obj);
+  // If obj is dead, only null-path is taken.
+  if (stopped())  return true;
```
```suggestion
  if (stopped()) { 
    return true;
  }
```

vnkozlov replied:
done

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5502,29 +5502,44 @@ bool LibraryCallKit::inline_array_partition() {
+  }
+  // get the address of the array
+  const TypeAryPtr* obj_t = _gvn.type(obj)->isa_aryptr();
+  if (obj_t == nullptr || obj_t->elem() == Type::BOTTOM ) {
```
```suggestion
  if (obj_t == nullptr || obj_t->elem() == Type::BOTTOM) {
```

vnkozlov replied:
done

chhagedorn reviewed (approved):
Only some style comments, otherwise, it looks good to me, too.

Were you able to reproduce this in mainline as well?

vnkozlov commented:
Moved common checks into new `check_array_sort_arguments()` method.
Moved intrinsic's stub presence check to the beginning of methods  and added assert to check signature size (number of stack slots) similar to other intrinsics.
I kept `null_check()` together with `stopped()` check for clarity.

Tested locally `java/util/Arrays/Sorting.java` on AVX512 machine to make sure intrinsics are generated and working.
Running tier1-3,stress,xcomp again for this update.

chhagedorn commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5489,52 +5489,68 @@ void LibraryCallKit::create_new_uncommon_trap(CallStaticJavaNode* uncommon_trap_
-
+// Common checks for array sorting intrinsics arguments.
+// Returns `true` if checks passed.
+bool LibraryCallKit::check_array_sort_arguments(Node* elementType, Node* obj, BasicType* bt) {
```
Thanks for doing the update, looks good! Just one small detail: You could use a reference here instead of a pointer. Then you can pass `bt` in and modify it directly.

vnkozlov replied:
Done.

vnkozlov commented:
/integrate
